File: 85-maximal-rectangle/maximal-rectangle.java
class Solution {
    public int maximalRectangle(char[][] matrix) {

        int rows=matrix.length;
        int cols=matrix[0].length;
        int[] heights=new int[cols];
        int maxArea=0;
        for(int i=0;i<rows;i++){
            for(int j=0;j<cols;j++){
                if(matrix[i][j]=='1'){
                    heights[j]++;
                }else{
                    heights[j]=0;
                }


            }
            int[] stack=new int[cols+1];
            int top=-1;
            for(int j=0;j<=cols;j++){
                int currHeight=(j==cols)?0:heights[j];
                while(top>=0&& heights[stack[top]]>=currHeight){
                    int h=heights[stack[top--]];
                    int width=(top<0)?j:j-stack[top]-1;
                    maxArea=Math.max(maxArea,h*width);
                }
                stack[++top]=j;
            }
        }
        
        return maxArea;
    }
}

File: 85-maximal-rectangle/maximal-rectangle.py
from typing import List

class Solution:
    def maximalRectangle(self, matrix: List[List[str]]) -> int:
        rows=len(matrix)
        cols=len(matrix[0])
        heights=[0]*cols
        max_area=0

        for i in range(rows):
            for j in range(cols):
                if matrix[i][j]=='1':
                    heights[j]+=1
                else:
                    heights[j]=0
            stack=[]

            for j in range(cols+1):
                curr_height=0 if j==cols else heights[j]
                while stack and heights[stack[-1]]>=curr_height:
                    h=heights[stack.pop()]
                    width=(j if not stack else j-stack[-1]-1)
                    max_area=max(max_area,h*width)
                stack.append(j)

        return max_area

            
        